assets/model_monitoring/components/tools/spec_version_upgrader.py

```diff
@@ -0,0 +1,173 @@
+        if spec.references:
+            for ref in spec.references:
+                yaml_content = re.sub(
+                    rf"(\s+)component: azureml://registries/azureml/components/{ref}/versions/\d+\.\d+\.\d+$",
+                    rf"\1component: azureml://registries/azureml/components/{ref}/versions/{spec.references[ref]}",
+                    yaml_content, flags=re.MULTILINE)
+
+        with open(spec.path, 'w') as f:
+            f.write(yaml_content)
+
+    def _upgrade_versions(self, specs: List[str]):
+        """Virtually upgrade version in spec_dict."""
+        upgraded_specs = set()
+        while specs:
+            spec_name = specs.pop()
+            if spec_name in upgraded_specs:
+                # skip if spec already upgraded, this can happen if one subgraph refer multiple components
+                continue
+            parent_specs = self._upgrade_version_and_reference(spec_name)
+            upgraded_specs.add(spec_name)
+            specs.extend(parent_specs)
+
+    def _upgrade_version_and_reference(self, spec_name: str) -> List[str]:
+        """
+        Upgrade spec version.
+
+        Also for each of its parent specs(spec that has reference to this spec), update the reference version to be 
+        new version. Return all the parent specs that refer the upgraded spec.
+        """
+        # find the spec for spec_name, upgrade its version+1
+        curr_spec = self._spec_dict.get(spec_name)
+        if not curr_spec:
+            raise ValueError(f"Spec {spec_name} not found")
+        v = curr_spec.version
+        curr_spec.version = Version(f"{v.major}.{v.minor}.{v.micro + 1}")
+        curr_spec.need_to_update = True
+
+        # upgrade reference version to the new version
+        parent_specs = []
+        for spec in self._spec_dict.values():
+            if spec.references and spec.references.get(spec_name):
+                spec.references[spec_name] = curr_spec.version
+                spec.need_to_update = True
+                parent_specs.append(spec.name)
+
+        return parent_specs
+
+    @staticmethod
+    def _scan_spec_folder(spec_folder: str) -> dict:
+        """Scan spec folder, return a dict of spec name to spec."""
+        spec_dict = {}
+        for root, _, files in os.walk(spec_folder):
+            for file_name in files:
+                if file_name == "spec.yaml":
+                    spec = Spec.from_yaml(os.path.join(root, file_name))
+                    spec_dict[spec.name] = spec
+
+        return spec_dict
+
+
+if __name__ == "__main__":
+    # TODO get specs from arguments
+    argparser = ArgumentParser()
+    argparser.add_argument("--specs", help="spec names to upgrade")
+    args = argparser.parse_args()
+    specs = args.specs.split(",") if args.specs else []
+    spec_folder = os.path.abspath(f"{os.path.dirname(__file__)}/..")
+    spec_version_upgrader = SpecVersionUpgrader(spec_folder)
+    spec_version_upgrader.upgrade_versions(specs)
```

Comment:
so we need to specify the component name(s)? If the name is not specified, it will upgrade all the components? What if the common utils is changed but not component itself changed? Can you add a .md file or add a comment with examples with usage?

Reply:
this tool figure out the subgraph/component dependency/reference, but you need to provide the first level component list.
I can make it to upgrade all component versions if no component provided.